Dataloader/Dataloader.py

```diff
@@ -35,78 +37,73 @@ def __getitem__(self, id):
         svs_path = self.tile_dataset['SVS_PATH'].iloc[id]
         svs_file = openslide.open_slide(svs_path)
         data = svs_file.read_region([self.tile_dataset["coords_x"].iloc[id], self.tile_dataset["coords_y"].iloc[id]], self.vis, self.dim).convert("RGB")
-        
-        for transform_step in self.transform.transforms:
-            if(isinstance(transform_step,ColourNorm.Macenko)):
-                HE, maxC = ColourNorm.Macenko().find_HE(data, get_maxC=True)
-                #data     = transform_step(data, self.tile_dataset['HE'].iloc[id])
-                data     = transform_step(data, HE, maxC)
-                continue
-            else:
-                data = transform_step(data)
 
-        ## Transform - Data Augmentation
-        #if self.transform: data = self.transform(data)
+        # Transform - Data Augmentation
+        if self.transform:
+            data = self.transform(data)
 
         if self.inference:
             return data
-
-        else: ## Training
+        else:
             label = self.tile_dataset[self.target].iloc[id]
-            if self.target_transform: label = self.target_transform(label)
+            if self.target_transform:
+                label = self.target_transform(label)
 
             return data, label
 
 
 class DataModule(LightningDataModule):
 
     def __init__(self, tile_dataset, train_transform=None, val_transform=None, batch_size=8, n_per_sample=np.Inf,
-                 train_size=0.7, test_size=0.15, target=None,
+                 test_size=0.15, val_size=0.15, target=None,
                  label_encoder=None, **kwargs):
         super().__init__()
 
         self.batch_size = batch_size
-        if(label_encoder):  tile_dataset[target] = label_encoder.transform(tile_dataset[target])  ## For classif only
+        self.num_workers = int(.8 * mp.Pool()._processes)  # number of workers for dataloader is 80% of maximum workers.
 
+        if label_encoder:
+            tile_dataset[target] = label_encoder.transform(tile_dataset[target])  # For classif only
 
-        # Split Test with differents WSI
-        # Code grabs test_size % of all data for testing. Then the remaining is split as
-        gss                       = GroupShuffleSplit(n_splits=1, test_size=test_size, random_state=42)
-        train_val_inds, test_inds = next(gss.split(tile_dataset, groups=tile_dataset['id_external']))
-        tile_dataset_train_val    = tile_dataset.iloc[train_val_inds].reset_index()
-        tile_dataset_test         = tile_dataset.iloc[test_inds].reset_index()
+        tile_dataset['id_external'] = tile_dataset['id_external'].apply(str)
 
-        # Select only N per sample from each class
-        tile_dataset_train_val   = tile_dataset_train_val.groupby(target).head(n=n_per_sample)
-        tile_dataset_test        = tile_dataset_test.groupby(target).head(n=n_per_sample)
+        tile_dataset_sampled = sampling_schemes.sample_N_per_WSI(tile_dataset, n_per_sample=n_per_sample)
```

Comment:
What's the problem with the groupby?

Reply:
The sampling method created issues for the validation set for some reason, this needs to be investigated. Reverted back to the previous sampling method for now.